### PR TITLE
Added `BeOneOf` for object comparisons with custom comparer support

### PR DESCRIPTION
**Pull request by xmarshal**
Added BeOneOf for object comparisons with custom comparer support
provided api for #2112

### CODE REVIEW

**Comment by IT-VBFK** [comment]
Hmm.. I think this should be discussed in an issue first (aka api review process ;) )

**Comment by xmarshal** [comment]
Api proposal added https://github.com/fluentassertions/fluentassertions/issues/2112

**Comment by IT-VBFK** [comment]
@xmarshal Do you mind covering your latest changes with tests?

File: Src/FluentAssertions/Primitives/ObjectAssertions.cs [hunk 1]
```diff
@@ -16,6 +16,94 @@ public ObjectAssertions(object value)
         : base(value)
     {
     }
+
+    /// <summary>
+    /// Asserts that a <typeparamref name="TExpectation"/> equals another <typeparamref name="TExpectation"/> using its <see cref="object.Equals(object)" /> implementation.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/ObjectAssertions.cs:21]: 🔧 You're not using `object.Equals` here. You're using `IEqualityComparer<T>`.

File: Src/FluentAssertions/Primitives/ObjectAssertions.cs [hunk 2]
```diff
@@ -16,6 +16,94 @@ public ObjectAssertions(object value)
...
+    /// </param>
+    public AndConstraint<ObjectAssertions> Be<TExpectation>(TExpectation expected, IEqualityComparer<TExpectation> comparer, string because = "", params object[] becauseArgs)
+    {
+        Guard.ThrowIfArgumentIsNull(comparer);
+
+        Execute.Assertion
+            .BecauseOf(because, becauseArgs)
+            .ForCondition(Subject is TExpectation subject && comparer.Equals(subject, expected))
+            .WithDefaultIdentifier(Identifier)
+            .FailWith("Expected {context} to be {0}{reason}, but found {1}.", expected,
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/ObjectAssertions.cs:41]: 🤔 I would also expect some reference to the comparer, should I not?

File: Tests/FluentAssertions.Specs/Primitives/ObjectAssertionSpecs.cs [hunk 3]
```diff
@@ -1290,3 +1364,51 @@ public void Dispose()
         // Ignore
     }
 }
+
+internal class SomeClass
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/ObjectAssertionSpecs.cs:1368]: 🔧 We prefer to keep the test classes as simple and specific for the individual test cases as possible, and also next to them (in the same grouping class, e.g. `BeOneOf`. See also

File: docs/_pages/basicassertions.md [hunk 4]
```diff
@@ -41,6 +41,12 @@ To assert that an object is equal to one of the provided objects, you can use
 theObject.Should().BeOneOf(obj1, obj2, obj3);
 ```
 
+To assert that an object is equal to one of the provided objects with custom equality comparer, you can use
+
+```csharp
+theObject.Should().BeOneOf<TObj>(new[] { obj1, obj2, obj3 } , new ObjEqualityCompmarer());
```
Comment (dennisdoomen) [inline docs/_pages/basicassertions.md:47]: 🤔  And what about the new `Be` and `NotBe` overloads?

File: docs/_pages/releases.md [hunk 5]
```diff
@@ -10,6 +10,7 @@ sidebar:
 ## Unreleased
 
 ### What's new
+* Added `BeOneOf` for object comparisons with custom comparer support -[#2111](https://github.com/fluentassertions/fluentassertions/pull/2111)
```
Comment (dennisdoomen) [inline docs/_pages/releases.md:13]: 🤔  And what about the new `Be` and `NotBe` overloads?

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Tests/FluentAssertions.Specs/Primitives/ObjectAssertionSpecs.cs [hunk 6]
```diff
@@ -189,6 +190,79 @@ public void When_a_value_is_one_of_the_specified_values_it_should_succeed()
...
+            var value = new SomeClass(3);
+
+            // Act
+            Action act = () => value.Should().BeOneOf(new[] { new SomeClass(4), new SomeClass(5) }, "because those are the valid values");
+
+            // Assert
+            act
+                .Should().Throw<XunitException>()
+                .WithMessage("Expected value to be one of {SomeClass(4), SomeClass(5)} because those are the valid values, but found SomeClass(3).");
+        }
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Primitives/ObjectAssertionSpecs.cs:222]: You can merge these two tests, there's no reason to have two tests here as they're testing a single _concept_.
I have written myself a note to clean this up.

The same comment applies for the two tests below.

**Comment by jnyrup** [comment]
@xmarshal Do you plan continuing this or anything blocking you?

File: Src/FluentAssertions/Primitives/ObjectAssertions.cs (hunk 2 above)
Comment (jnyrup) [reply Src/FluentAssertions/Primitives/ObjectAssertions.cs:41]: You mean something like including `comparer.GetType()`?

We do something similar when using an `IEqualityComparer<T>` in equivalency assertions.
https://github.com/fluentassertions/fluentassertions/pull/1284/files#diff-f060184f7198367f57bfcf2bbfc8ce3ddc5ce5db1949c93e376d6d3c780c53b8R37

**Comment by jnyrup** [comment]
@xmarshal checking in one more time to see if you'd like to drive this PR to completion.
I really like the addition, so if you're no longer interested in or have the time to finish this, I'll pick it up and finish the last bits.

**Pull request closed without merging**